Cookbook/Proposals/SmarterDiffs.md

```diff
@@ -0,0 +1,97 @@
+At the same time if a single state assert fails it's easier to understand what failed as we are comparing less values at a time.
+
+#### Option 2 - assert all states at once
+
+Example:
+
+```swift
+expect(states).to(equal([
+    makeState(status: .idle),
+    makeState(status: .loading),
+    makeState(
+        shouldShowNHSBanner: true,
+        shouldShowHealthcheckBanner: true,
+        shouldShowTestAndKitsBanner: true,
+        shouldShowGPAtHandStatusCard: true,
+        status: .loaded
+    )
+]))
+```
+
+This eliminates the issue of the first approach when some state changes may be ignored by the test. It comes with a cost of actually creating all the intermediate states that might not be directly related to what test is testing which adds some amount of boilerplate. Another issue is that when this single assert fails it's very hard to extract what actually cuased this failure from the error message until we go and check each states separately. First option also suffers from this issue when each state consists of a lot of properties but it becomes much worse with a single assert.
+
+## Motivation
+
+As demonstrated our current approaches suffer from few issues which result in not ideal developer experience when it comes to debugging test failures.
+
+## Proposed solution
+
+To improve the experience with debugging test failures we propose to use a smarter diffing approach as in [this library](https://github.com/krzysztofzablocki/Difference) by Krzysztof Zabłocki. The idea is that when comparing two instances we will be using `Mirror` type and `dump` function to first get stored properties and then get their textual representation and this way combine a in-memory snapshot of the value. This way we can recursively compare each property and only include those that differ in the final failure message. 
+
+The fact that `Mirror` only provides access to stored properties is not an issue as computed properties are products of stored properties (or in case of enums of the enum cases and thier associated values) and for same stored properties values should return the same result.
+
+Example:
+
+```swift
+// equalDiff is a name suggested in the Diffable repo README, we can choose a different one
+expect(states).to(equalDiff([
+    makeState(status: .idle),
+    makeState(status: .loading),
+    makeState(
+        shouldShowNHSBanner: false, // this value causes failure
+        shouldShowHealthcheckBanner: true,
+        shouldShowTestAndKitsBanner: true,
+        shouldShowGPAtHandStatusCard: true,
+        status: .loaded
+    )
+]))
+```
```

Comment:
I am not convinced that writing all this boiler plate is worth it. If my test is checking a particular property in the state after an action of steps, I would like the test to be focussed on those subset of properties. It makes the reading and understanding the purpose of the test easier.
As to the problem of crashes when states array's index is out of bounds, maybe we can use guard statement
e.g. 
```
guard states.count > 10 else { XCTFail()} 
expect(states.first?.lastName).to(beNil())
```

Reply:
There are benefits and tradeoffs in each of these options but regardless which one we use they both benefit from the proposal

Reply:
@chitrakotwani But if you only assert on the subset of properties in question, then you will miss catching the corruption to state of properties outside the subset. Its best to assert the entire state equals your expected state. You could differentiate the setup of the expectation however with only the subset of properties (to understand what you are concerned about in the specific test case).

Reply:
@anilputtabuddhi - This would make the tests more brittle because an unexpected change in one state property would break quite a few tests at once. This would make it difficult to pinpoint the blame of breaking tests.

But as @ilyapuchka  suggests, this is probably out of scope of discussion of this proposal anyway.

Reply:
Agree with @anilputtabuddhi. We will always compromise between coverage and clarity of tests. I believe we should prefer coverage but have a clarity in mind (test should be steal readable and understandable). That's not what the proposal is about though =)

Reply:
@chitrakotwani The tests would not be called brittle in the scenario you explained - An unexpected state change should cause failures and thats what tests are for. A brittle test is one that fails and succeeds intermittently on different runs. It is also possible to architect tests with proper fixture designs that help in test setup to be able to pinpoint breakages as you say.

Reply:
@anilputtabuddhi - Brittle tests, as name implies, are the ones which break easily. If a test will begin to depend on too many state's properties, they will start breaking easily. A small change somewhere will have unintended effects in tests in other areas. 
Tests which fail and succeed intermittently on different runs could be called unreliable?